core/src/main/java/com/datastrato/gravitino/lock/LockManager.java

```diff
@@ -0,0 +1,126 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * LockManager is a lock manager that manages the tree locks. It will service as a factory to create
+ * the tree lock and do the cleanup for the stale tree locks node.
+ */
+public class LockManager {
+  private static final Logger LOG = LoggerFactory.getLogger(LockManager.class);
+  @VisibleForTesting final TreeLockNode treeLockRootNode;
+  final AtomicLong totalNodeCount = new AtomicLong(1);
+
+  // TODO (yuqi) make these two configurable
+  static final long MAX_TREE_NODE_IN_MEMORY = 10000L;
+  // If the total node count is less than this value, we will not do the cleanup.
+  @VisibleForTesting static final long MIN_TREE_NODE_IN_MEMORY = 1000L;
+
+  private final ScheduledThreadPoolExecutor lockCleaner;
+
+  public LockManager() {
+    treeLockRootNode = new TreeLockNode(NameIdentifier.ofRoot(), this);
+    this.lockCleaner =
+        new ScheduledThreadPoolExecutor(
+            1,
+            new ThreadFactoryBuilder()
+                .setDaemon(true)
+                .setNameFormat("tree-lock-cleaner-%d")
+                .build());
+
+    lockCleaner.scheduleAtFixedRate(
+        () -> {
+          long nodeCount = totalNodeCount.get();
+          LOG.info("Total tree lock node count: {}", nodeCount);
+          if (nodeCount > MAX_TREE_NODE_IN_MEMORY) {
+            treeLockRootNode
+                .getAllChildren()
+                .forEach(child -> evictStaleNodes(child, treeLockRootNode));
+          }
+        },
+        120,
+        5,
+        TimeUnit.SECONDS);
+  }
+
+  /**
+   * Evict the stale nodes from the tree lock node.
+   *
+   * @param treeNode The tree lock node to evict.
+   * @param parent The parent of the tree lock node.
+   */
+  @VisibleForTesting
+  void evictStaleNodes(TreeLockNode treeNode, TreeLockNode parent) {
+    // We will not evict the root node if the total node count is less than the
+    // MIN_TREE_NODE_IN_MEMORY.
+    // Do not need to consider thread-safe issues.
+    if (totalNodeCount.get() < MIN_TREE_NODE_IN_MEMORY) {
+      return;
+    }
+
+    // Handle from leaf nodes first.
+    treeNode.getAllChildren().forEach(child -> evictStaleNodes(child, treeNode));
+
+    // Handle self node.
+    if (treeNode.getReferenceCount() == 0) {
+      synchronized (parent) {
+        // Once goes here, the parent node has been locked, so the reference could not be changed.
+        if (treeNode.getReferenceCount() == 0) {
+          parent.removeChild(treeNode.getIdent());
+          totalNodeCount.decrementAndGet();
+          LOG.info("Evict stale tree lock node '{}' and all its children", treeNode.getIdent());
+        } else {
+          treeNode.getAllChildren().forEach(child -> evictStaleNodes(child, treeNode));
+        }
+      }
+    }
+  }
+
+  /**
+   * Create a tree lock with the given identifier.
+   *
+   * @param identifier The identifier of the tree lock.
+   * @return The created tree lock.
+   */
```

Comment:
Why some method you add the javadoc, some not?

Reply:
Fixed